tools/e2etesting/TestEventProcessor/TestEventProcessor.Businesslogic/EventProcessorWrapper.cs

```diff
@@ -0,0 +1,254 @@
+                _config.EventHubConsumerGroup,
+                _config.IoTHubEventHubEndpointConnectionString
+            );
+
+            var partitions = await eventHubConsumerClient
+                .GetPartitionIdsAsync(ct)
+                .ConfigureAwait(false);
+            _initializedPartitions = partitions.ToDictionary(item => item, _ => false);
+            _lockInitializedPartitions = new SemaphoreSlim(1, 1);
+
+            _logger.LogInformation("Connecting to IoT Hub...");
+            _client = new EventProcessorClient(
+                blobContainerClient,
+                _config.EventHubConsumerGroup,
+                _config.IoTHubEventHubEndpointConnectionString
+            );
+
+            _client.PartitionClosingAsync += Client_PartitionClosingAsync;
+            _client.PartitionInitializingAsync += Client_PartitionInitializingAsync;
+            _client.ProcessEventAsync += Client_ProcessEventAsync;
+            _client.ProcessErrorAsync += Client_ProcessErrorAsync;
+        }
+
+        /// <summary>
+        /// Start processing of events. If the method was called previously, then it will only
+        /// re-enable processing.
+        /// </summary>
+        public async Task StartProcessingAsync(CancellationToken ct) {
+            if (_client == null) {
+                throw new InvalidOperationException("EventProcessorWrapper has not been initialized.");
+            }
+
+            // If processing is active for all partitions then we only need to enable message propagation.
+            _lockInitializedPartitions.Wait();
+            try {
+                if (_initializedPartitions.Count(kvp => kvp.Value) == _initializedPartitions.Count) {
+                    _logger.LogInformation("Enabling monitoring of events...");
+                    _monitoringEnabled = true;
+                    return;
+                }
+            }
+            finally {
+                _lockInitializedPartitions.Release();
+            }
+
+            if (!_client.IsRunning) {
+                _logger.LogInformation("Starting monitoring of events...");
+                await _client.StartProcessingAsync(ct).ConfigureAwait(false);
+            }
+
+            await WaitForPartitionInitialization(ct).ConfigureAwait(false);
+
+            _monitoringEnabled = true;
+        }
+
+        /// <summary>
+        /// Stop propagation of messages to downstream event handler.
+        /// This will not close connection to EventHub.
+        /// </summary>
+        public void StopProcessing() {
+            _logger.LogInformation("Disabling monitoring of events...");
+            _monitoringEnabled = false;
+        }
+
+        /// <summary>
+        /// Wait until we receive confirmation that monitoring of each partition has started.
+        /// </summary>
+        private async Task WaitForPartitionInitialization(CancellationToken ct) {
+            var sw = Stopwatch.StartNew();
```

Comment:
>Stopwatch [](http://example.com/codeflow?start=21&length=9)

shouldn't we stop this on the finally block?

Reply:
No, not really. `Stop()` just stops tracking the elapsed time. That will happen anyway since the object will be destroyed soon after the `finally` block.